File: 2021/09/17/main.py
class Solution:
    def intersect(self, nums1: List[int], nums2: List[int]) -> List[int]:
        m1, m2, r = {}, {}, []
        for n in nums1:
            if n in m1:
                m1[n] += 1
            else:
                m1[n] = 1
        for n in nums2:
            if n in m2:
                m2[n] += 1
            else:
                m2[n] = 1
        
        if len(list(m1)) > len(list(m2)):
            m1, m2 = m2, m1

        for n in m1:
            if n in m2:
                m = min(m1[n], m2[n])
                for i in range(m):
                    r.append(n)
        return r
        

File: 2021/09/17/main.go
package main

func intersect(nums1 []int, nums2 []int) []int {
	res := []int{}
	m1 := getInts(nums1)
	m2 := getInts(nums2)

	if len(m1) > len(m2) {
		m1, m2 = m2, m1
	}
	for n := range m1 {
		m1[n] = min(m1[n], m2[n])
	}

	for n, size := range m1 {
		for i := 0; i < size; i++ {
			res = append(res, n)
		}
	}

	return res
}

func getInts(a []int) map[int]int {
	res := make(map[int]int, len(a))

	for i := range a {
		res[a[i]]++
	}

	return res
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}
